pkg/kademlia/queue.go

```diff
@@ -4,57 +4,102 @@
 package kademlia
 
 import (
+	"container/heap"
 	"math/big"
 
+	"storj.io/storj/pkg/dht"
 	"storj.io/storj/pkg/pb"
 )
 
-// An Item is something we manage in a priority queue.
-type Item struct {
+//XorQueue is a priority queue where the priority is key XOR distance
+type XorQueue struct {
+	maxLen int
+	items  items
+}
+
+//NewXorQueue returns a items with priority based on XOR from targetBytes
+func NewXorQueue(size int) *XorQueue {
+	return &XorQueue{items: make(items, 0, size), maxLen: size}
+}
+
+//Insert adds Node onto the queue
+func (x *XorQueue) Insert(target dht.NodeID, nodes []*pb.Node) {
+	targetBytes := new(big.Int).SetBytes(target.Bytes())
+	//insert new nodes
+	for _, node := range nodes {
+		heap.Push(&x.items, &item{
+			value:    node,
+			priority: new(big.Int).Xor(targetBytes, new(big.Int).SetBytes([]byte(node.GetId()))),
+		})
+	}
+	//resize down if we grew too big
+	if x.items.Len() > x.maxLen {
+		olditems := x.items
+		x.items = items{}
+		for i := 0; i < x.maxLen && len(olditems) > 0; i++ {
+			item := heap.Pop(&olditems)
+			heap.Push(&x.items, item)
+		}
+		heap.Init(&x.items)
```

Comment:
Note there are neat data-structures for this https://en.wikipedia.org/wiki/Double-ended_priority_queue, but we can do them in a later PR when necessary.